{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "powershell"
    }
   },
   "outputs": [],
   "source": [
    "#Bash script - processing of wastewater sequencing raw_data\n",
    "\n",
    "#!/bin/bash\n",
    "#python sanitize_bed.py \"${refs}\"swift_primersv2.bed \n",
    "#python prepare_amplicon_info.py \"${refs}\"swift_primersv2.bed \"${refs}\"swift_primersv2_info.tsv \"${refs}\"swift_primersv2_info.tsv\n",
    "\n",
    "module load samtools/1.15\n",
    "module load bwa/0.7.17\n",
    "\n",
    "module load vcflib/1.0.1\n",
    "module load snpEff/5.0\n",
    "\n",
    "batch=$2\n",
    "refs=\"/gpfs/gsfs12/users/Irp-jiang/share/covid_data/WWBE/info/\"\n",
    "base=\"/gpfs/gsfs12/users/Irp-jiang/share/covid_data/WWBE/${batch}/\"\n",
    "\n",
    "#mkdir \"${base}analysis_galaxy/\"\n",
    "analysis=\"${base}analysis_galaxy/\"\n",
    "#mkdir \"${analysis}tsvfiles\"\n",
    "#mkdir \"${analysis}vcffiles\"\n",
    "#mkdir \"${analysis}merged_file\"\n",
    "#mkdir \"${analysis}bamfiles\"\n",
    "\n",
    "\n",
    "#aligning\n",
    "bwa mem -t 32 -v 1 \"${refs}\"NC_045512.2.fasta \"${base}raw_data/${1}\"_1.fastq.gz \"${base}raw_data/${1}\"_2.fastq.gz | samtools view -@ 32 -b -f 1 -F 268 -q 20 -s 1.0 | samtools sort -@ 32 -o \"${analysis}bamfiles/${1}\".sorted.bam\n",
    "echo \"Sample was initially aligned with BWA\"\n",
    "\n",
    "#trimming primers and base quality\n",
    "module load ivar/1.3.1\n",
    "ivar trim -e -m 1 -q 0 -b \"${refs}\"swift_primersv2.bed -p \"${analysis}bamfiles/${1}\".trimmed -i \"${analysis}bamfiles/${1}\".sorted.bam\n",
    "echo \"Amplification primers were removed\"\n",
    "\n",
    "#lofreq realign - may need to separate samtools to controls the version\n",
    "module load lofreq/2.1.5\n",
    "lofreq viterbi --defqual 2 -f \"${refs}\"NC_045512.2.fasta \"${analysis}bamfiles/${1}\".trimmed.bam | samtools sort -@ 32 -o \"${analysis}bamfiles/${1}\".trimmed.realigned.sorted.bam\n",
    "echo \"Bam file was re-aligned using lofreq\"\n",
    "\n",
    "#lofreq indelquality and indexing\n",
    "lofreq indelqual --dindel -f \"${refs}\"NC_045512.2.fasta -o \"${analysis}bamfiles/${1}\".trimmed.realigned.indelqual.bam \"${analysis}bamfiles/${1}\".trimmed.realigned.sorted.bam \n",
    "echo \"Bam file annotated with indel information\"\n",
    "\n",
    "#lofreq variant call #1\n",
    "lofreq call --min-cov 5 --max-depth 1000000 -q 30 -Q 30 -e --min-mq 20 --sig 0.0005 --bonf dynamic --no-default-filter --no-default-filter -f \"${refs}\"NC_045512.2.fasta --call-indels -o \"${analysis}vcffiles/${1}\"_trimmed.vcf \"${analysis}bamfiles/${1}\".trimmed.realigned.indelqual.bam\n",
    "echo \"First round of variant call with lofreq finished\"\n",
    "\n",
    "#lofreq call --min-cov 20  -q 20 -Q 20 -e --min-mq 20 -f \"${refs}\"NC_045512.2.fasta --call-indels -o \"${analysis}vcffiles/${1}\"_trimmed.vcf \"${analysis}vcffiles/${1}\".trimmed.realigned.indelqual.bam\n",
    "lofreq filter -V 0 -v 5 -a 0.05 -A 0.95 -i \"${analysis}vcffiles/${1}\"_trimmed.vcf -o \"${analysis}vcffiles/${1}\"_trimmed_filtered.vcf\n",
    "echo \"Fisrt round of variant call filtered\"\n",
    "\n",
    "#step to revome primer bias reads\n",
    "module load ivar/1.3.1\n",
    "ivar getmasked -i \"${analysis}vcffiles/${1}\"_trimmed_filtered.vcf -b \"${refs}\"swift_primersv2.bed -f \"${refs}\"swift_primersv2_info.tsv -p \"${analysis}${1}\"_primer_mismatchers_indices\n",
    "python completemask.py \"${analysis}${1}\"_primer_mismatchers_indices.txt \"${refs}\"swift_primersv2_info.tsv \"${analysis}${1}\"_primer_mismatchers_indices_v2.txt\n",
    "ivar removereads -i \"${analysis}bamfiles/${1}\".trimmed.realigned.indelqual.bam -b \"${refs}\"swift_primersv2.bed -t \"${analysis}${1}\"_primer_mismatchers_indices_v2.txt -p \"${analysis}bamfiles/${1}\".trimmed.realigned.indelqual.readsrem.bam\n",
    "echo \"Bias reads removed with ivar\"\n",
    "\n",
    "#lofreq variant call #2\n",
    "module load lofreq/2.1.5\n",
    "#lofreq call --min-cov 20  -q 20 -Q 20 -e --min-mq 20 -f \"${refs}\"NC_045512.2.fasta --call-indels -o \"${analysis}vcffiles/${1}\"_trimmed_v2.vcf \"${analysis}vcffiles/${1}\".trimmed.realigned.indelqual.readsrem.bam\n",
    "lofreq call --min-cov 5 --max-depth 1000000 -q 30 -Q 30 -e --min-mq 20 --sig 0.0005 --bonf dynamic --no-default-filter --no-default-filter -f \"${refs}\"NC_045512.2.fasta --call-indels -o \"${analysis}vcffiles/${1}\"_trimmed_v2.vcf \"${analysis}bamfiles/${1}\".trimmed.realigned.indelqual.readsrem.bam\n",
    "echo \"Second round of variant call with lofreq finished - using removed reads bam\"\n",
    "\n",
    "#vcf files intersect #1\n",
    "vcfintersect -r \"${refs}\"NC_045512.2.fasta -v -w 0  -i \"${analysis}vcffiles/${1}\"_trimmed_v2.vcf \"${analysis}vcffiles/${1}\"_trimmed_filtered.vcf > \"${analysis}vcffiles/${1}\"_trimmed_intersect.vcf\n",
    "echo \"VCF_trimmed_v2 intersected with vcf_trimmed_filtered\"\n",
    "\n",
    "#re-naming\n",
    "sed -r --sandbox -e 's/^(#CHROM.+)$/##FILTER=<ID=AmpliconRemoval,Description=\"Variant removed upon removal of amplicon\">\\n\\1/g' -e 's/(.+\\t)PASS(\\t.+)/\\1AmpliconRemoval\\2/g' \"${analysis}vcffiles/${1}\"_trimmed_intersect.vcf > \"${analysis}vcffiles/${1}\"_trimmed_renamed.vcf\n",
    "echo \"Intersected vcf re-named\"\n",
    "\n",
    "#vcf files union\n",
    "vcfintersect -r \"${refs}\"NC_045512.2.fasta -w 0 -u \"${analysis}vcffiles/${1}\"_trimmed_renamed.vcf  \"${analysis}vcffiles/${1}\"_trimmed_v2.vcf > \"${analysis}vcffiles/${1}\"_trimmed_union.vcf\n",
    "lofreq filter -V 0 -v 0 -a 0.0 -A 0.0 -b fdr -c 0.001 --print-all -i \"${analysis}vcffiles/${1}\"_trimmed_union.vcf -o \"${analysis}vcffiles/${1}\"_trimmed_union_filtered.vcf\n",
    "echo \"Final vcf created - union of vcf_trimmed_renamed with vcf_trimmed_v2\"\n",
    "\n",
    "#annotation\n",
    "#wget https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_NC_045512.2.zip\n",
    "#unzip snpEff_v5_0_NC_045512.2.zip will create data folder\n",
    "snpEff eff -nodownload -dataDir ${refs}data -i vcf -o vcf -formatEff -classic -no-downstream -no-intergenic -no-upstream -ud 0 -stats stats.html -noLog NC_045512.2 \"${analysis}vcffiles/${1}\"_trimmed_union_filtered.vcf > \"${analysis}vcffiles/${1}\"_trimmed_union_snpEff.vcf\n",
    "echo \"Final VCF annotated\"\n",
    "\n",
    "python \"${refs}\"add_variant_tsv_improved.py  \"${analysis}vcffiles/${1}\"_trimmed_union_snpEff.vcf  \"${analysis}vcffiles/${1}\"_trimmed_union_snpEff_final.tsv\n",
    "\n",
    "module load samtools/1.15\n",
    "samtools fastq -1 \"${analysis}fastqfiles/${1}\"_1.fq.gz -2 \"${analysis}fastqfiles/${1}\"_2.fq.gz -@ 32 -n \"${analysis}bamfiles/${1}\".trimmed.realigned.indelqual.bam\n",
    "echo \"Generated FASTQ files for SRA submission\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After processing of raw data - get metrics of the sequencing batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "powershell"
    }
   },
   "outputs": [],
   "source": [
    "#Get metrics of the sequencing batch\n",
    "\n",
    "module load samtools\n",
    "echo -e \"File \\tMean \\tBreadth \\tRead_count\" >> /data/salgadofontenr2/ww_cov_2021/metrics_galaxy_${1}.txt\n",
    "\n",
    "for file in /gpfs/gsfs12/users/Irp-jiang/share/covid_data/WWBE/${1}/analysis_galaxy/bamfiles/*.trimmed.realigned.indelqual.readsrem.bam;\n",
    "do\n",
    "\tprefix=$(echo $file | cut -d \"/\" -f 12 | cut -d \".\" -f 1)\n",
    "\tcount=$(samtools view -c -F 4 \"${file}\")\n",
    "\tmean=$(samtools depth -a \"${file}\" | awk '{c++;s+=$3}END{print s/c}')\n",
    "\tbreadth=$(samtools depth -a \"${file}\" | awk '{c++; if($3>10) total+=1}END{print (total/(c))*100}')\n",
    "\techo -e \"${prefix} \\t${mean} \\t${breadth} \\t${count}\" >> /data/salgadofontenr2/ww_cov_2021/metrics_galaxy_${1}.txt\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Python code to edit VCF file after raw data processing - add_variant_tsv_improved.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "import vcf\n",
    "\n",
    "GENE_MAP = {'ORF1a': [266, 13468],\n",
    "            'ORF1b': [13468, 21555],\n",
    "            'S': [21563, 25384],\n",
    "            'ORF3a': [25393, 26220],\n",
    "            'E': [26245, 26472],\n",
    "            'M': [26523, 27191],\n",
    "            'ORF6': [27202, 27387],\n",
    "            'ORF7a': [27394, 27759],\n",
    "            'ORF7b': [27756, 27887],\n",
    "            'ORF8': [27894, 28259],\n",
    "            'N': [28274, 29533],\n",
    "            'ORF10': [29558, 29674]}\n",
    "\n",
    "def read_lofreq(filename):\n",
    "    lofreq_calls = pd.DataFrame(columns=[\"REGION\", \"POS\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \n",
    "                                         \"REF_DP\", \"REF_RV\", \"ALT_DP\", \"ALT_RV\",\n",
    "                                         \"AF\", \"TOTAL_DP\", \"STRAND-BIAS\", \"EFFECT\"])\n",
    "    vcf_reader = vcf.Reader(filename=filename)\n",
    "    for row in vcf_reader:\n",
    "        if len(row.FILTER) == 0 and \"EFF\" not in row.INFO:\n",
    "            lofreq_calls = lofreq_calls.append({\"REGION\": row.CHROM, \n",
    "                                            \"POS\": int(row.POS),\n",
    "                                            \"REF\": str(row.REF),\n",
    "                                            \"ALT\": str(row.ALT[0]),\n",
    "                                            \"QUAL\": row.QUAL, \n",
    "                                            \"FILTER\": \"\",\n",
    "                                            \"REF_DP\": row.INFO[\"DP4\"][0] + row.INFO[\"DP4\"][1], \n",
    "                                            \"REF_RV\": row.INFO[\"DP4\"][1],\n",
    "                                            \"ALT_DP\": row.INFO[\"DP4\"][2] + row.INFO[\"DP4\"][3],\n",
    "                                            \"ALT_RV\": row.INFO[\"DP4\"][3],\n",
    "                                            \"AF\": row.INFO[\"AF\"],\n",
    "                                            \"TOTAL_DP\": row.INFO[\"DP\"],\n",
    "                                            \"STRAND-BIAS\":row.INFO[\"SB\"],\n",
    "                                            \"EFF\":\"\"\n",
    "                                           }, \n",
    "                                           ignore_index=True)\n",
    "        if len(row.FILTER) > 0 and \"EFF\" in row.INFO:\n",
    "            lofreq_calls = lofreq_calls.append({\"REGION\": row.CHROM, \n",
    "                                            \"POS\": int(row.POS),\n",
    "                                            \"REF\": str(row.REF),\n",
    "                                            \"ALT\": str(row.ALT[0]),\n",
    "                                            \"QUAL\": row.QUAL, \n",
    "                                            \"FILTER\": row.FILTER[0],\n",
    "                                            \"REF_DP\": row.INFO[\"DP4\"][0] + row.INFO[\"DP4\"][1], \n",
    "                                            \"REF_RV\": row.INFO[\"DP4\"][1],\n",
    "                                            \"ALT_DP\": row.INFO[\"DP4\"][2] + row.INFO[\"DP4\"][3],\n",
    "                                            \"ALT_RV\": row.INFO[\"DP4\"][3],\n",
    "                                            \"AF\": row.INFO[\"AF\"],\n",
    "                                            \"TOTAL_DP\": row.INFO[\"DP\"],\n",
    "                                            \"STRAND-BIAS\":row.INFO[\"SB\"],\n",
    "                                            \"EFF\":row.INFO[\"EFF\"]\n",
    "                                           }, \n",
    "                                           ignore_index=True)\n",
    "        if len(row.FILTER) == 0 and \"EFF\" in row.INFO:\n",
    "            lofreq_calls = lofreq_calls.append({\"REGION\": row.CHROM, \n",
    "                                            \"POS\": int(row.POS),\n",
    "                                            \"REF\": str(row.REF),\n",
    "                                            \"ALT\": str(row.ALT[0]),\n",
    "                                            \"QUAL\": row.QUAL, \n",
    "                                            \"FILTER\": \"\",\n",
    "                                            \"REF_DP\": row.INFO[\"DP4\"][0] + row.INFO[\"DP4\"][1], \n",
    "                                            \"REF_RV\": row.INFO[\"DP4\"][1],\n",
    "                                            \"ALT_DP\": row.INFO[\"DP4\"][2] + row.INFO[\"DP4\"][3],\n",
    "                                            \"ALT_RV\": row.INFO[\"DP4\"][3],\n",
    "                                            \"AF\": row.INFO[\"AF\"],\n",
    "                                            \"TOTAL_DP\": row.INFO[\"DP\"],\n",
    "                                            \"STRAND-BIAS\":row.INFO[\"SB\"],\n",
    "                                            \"EFF\":row.INFO[\"EFF\"],\n",
    "                                           }, \n",
    "                                           ignore_index=True)\n",
    "        if len(row.FILTER) > 0 and \"EFF\" not in row.INFO:\n",
    "            lofreq_calls = lofreq_calls.append({\"REGION\": row.CHROM, \n",
    "                                            \"POS\": int(row.POS),\n",
    "                                            \"REF\": str(row.REF),\n",
    "                                            \"ALT\": str(row.ALT[0]),\n",
    "                                            \"QUAL\": row.QUAL, \n",
    "                                            \"FILTER\": row.FILTER[0],\n",
    "                                            \"REF_DP\": row.INFO[\"DP4\"][0] + row.INFO[\"DP4\"][1], \n",
    "                                            \"REF_RV\": row.INFO[\"DP4\"][1],\n",
    "                                            \"ALT_DP\": row.INFO[\"DP4\"][2] + row.INFO[\"DP4\"][3],\n",
    "                                            \"ALT_RV\": row.INFO[\"DP4\"][3],\n",
    "                                            \"AF\": row.INFO[\"AF\"],\n",
    "                                            \"TOTAL_DP\": row.INFO[\"DP\"],\n",
    "                                            \"STRAND-BIAS\":row.INFO[\"SB\"],\n",
    "                                            \"EFF\":\"\"\n",
    "                                           }, \n",
    "                                           ignore_index=True)\n",
    "    if lofreq_calls.empty:\n",
    "        return lofreq_calls\n",
    "    lofreq_calls[\"Variant\"] = lofreq_calls.apply(lambda row:\n",
    "                                         str(row[\"POS\"]) + \":\" + \\\n",
    "                                         str(row[\"ALT\"][1:]) if len(row[\"ALT\"]) > 1 else \\\n",
    "                                         (str(row[\"POS\"]) + \"-\" + \\\n",
    "                                         str(row[\"POS\"] + len(row[\"REF\"][1:]) + 1) if len(row[\"REF\"]) > 1 else \\\n",
    "                                         str(row[\"REF\"]) + str(row[\"POS\"]) + \\\n",
    "                                         str(row[\"ALT\"])), axis=1)\n",
    "    lofreq_calls[\"EFFECT\"] = lofreq_calls.apply(lambda row:\n",
    "                                         str(row[\"EFF\"]).split(\"|\")[1] if row[\"EFF\"] else \\\n",
    "                                         \"SILENT\", axis=1)\n",
    "    lofreq_calls[\"AA_MUT\"] = lofreq_calls.apply(lambda row:\n",
    "                                         str(row[\"EFF\"]).split(\"|\")[3] if row[\"EFF\"] else \\\n",
    "                                         \"\", axis=1)\n",
    "    lofreq_calls[\"GENE\"] = lofreq_calls.apply(lambda row:\n",
    "                                         str(row[\"EFF\"]).split(\"|\")[5] if row[\"EFF\"] else \\\n",
    "                                         \"Non-coding\", axis=1)\n",
    "    #lofreq_calls = lofreq_calls[lofreq_calls[\"FILTER\"] != \"AmpliconRemoval\"]\n",
    "    return lofreq_calls\n",
    "\n",
    "def main():\n",
    "    file = sys.argv[1]\n",
    "    new_file = read_lofreq(file)\n",
    "    new_file.to_csv(sys.argv[2], sep=\"\\t\", index=False)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "397704579725e15f5c7cb49fe5f0341eb7531c82d19f2c29d197e8b64ab5776b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
